built foundation of a new decompression system

satorialist opened the pull request:
plz rate 10/10 and report any remainders of proprietary code so that I can fix thx

Commit:
restored old commits after forced reset

in order to get a clean working tree, compression now behaves less weirdly but lzmaread still fails

Commit:
pave the way for a new decompression system

the fundamental problem that is posed by our current streambase is as follows: for the compression, the streambase has to read from 1 file into the buffer, compress, and write to 2 files. For the decompression however, the streambase has to read from 2 files, decompress, and write to the buffer. The current way things are handled, which is, take a filepath, concatenate .meta and .data respectively and do your stuff, which is ideal for the compression, but not so much for the decompression, as I would like to open 2 separate streams for meta and data respectively. This new system should accomodate for that.

other things that changed:

- taken the bkyin and decomp from bkystreambase
- nice lil status messages for lzmaio which are now commented out
- NEVER PRINT BUFFERS TO CONSOLE

Commit:
experimental parsing of compressed sections

WIP

Commit:
compressed section parsing should now try to parse

of course a bug that makes the parser read for an unlimited amount of time still persists, but we can do something about it once we find out what the damn problem is

Commit:
removed massive memleaks

UgiNyaa committed:
lzmaio fix

Commit:
Merge pull request #2 from UgiNyaa/satorialist-master

lzmaio fix

Commit:
finalized compressed section reading, now reading metadata

some changes in blockymetadata and blockydecompression to prepare the parsing of metadata

Commit:
added blockynumbersaver

Commit:
some parsing restructures

Commit:
more structural stuff

Commit:
changed bitreader to use LZMAFILE

Commit:
metadata parsing using bitreader again

Commit:
implemented new fill_buffer system

which basically works as follows:

- the buffer filling process has two modes: fill with meta, and fill with data
- IF the mode is fill with meta, the decompressionparser will look two fill the buffer with chars from .meta UNTIL we hit the start of a compressed section, the parser switches mode
- IF the mdoe is fill with data, the decompressionparser will take the information of the current compressed section, and parse UNTIL it is fully parsed, the parser once again switches mode, although right now the decompression part isn't implemented yet

Commit:
lots o error checking

fill_buffer fails to fill the buffer when the buffersize is too small, which shouldn't happen
also, we got fancy std::cout prints now

UgiNyaa commented on src/blockynumbersaver.hpp:
```diff
@@ -0,0 +1,16 @@
+{
+private:
+    int index;
+    BlockyNumber* values;
```
maybe use vector here?

satorialist replied:
the array is later initalized with a fixed size

UgiNyaa replied:
You could use the [resize](http://www.cplusplus.com/reference/vector/vector/resize/) method of vector. Since vector automatically manages deallocation when it is out of scope, it would be safer to use.

satorialist replied:
I guess so, will do

UgiNyaa commented on src/decomp/decompstreambuffer.cpp:
```diff
@@ -0,0 +1,209 @@
+    if (!(mode & std::ios::in) || !opened)
+        return EOF;
+
+    int num = decompression->fill_buffer(buffer, bufferSize, sections);
```
how about giving "sections" vector to decompression via constructor, instead of having to hold the reference in decompstreambuffer?

satorialist replied:
good idea, will do

UgiNyaa reviewed (commented):
A review about little things I noticed

UgiNyaa commented:
How far did you come with decompression?

satorialist commented:
in this build fill_buffer switches between reading metadata and data going by the section sizes given in meta, it can read metadata but it returns 0 when reading data for now, actual decomp starts when bitreader, the decompression functions and the current fill_buffer switching were tested

Commit:
implemented rough structure of decompression data parsing

BlockyAlgorithm and BlockyDecompression need to be adjusted to the new system before it works

few other things:
- BlockyNumberSaver now uses a vector
- a reference to compressed sections is passed to DecompressionParser at initialization instead of per function call
- bufferSize -> buffer_size to adhere to convention

Commit:
blockydecompression now uses numbersaver correctly

Commit:
blocky compression hotfix

Commit:
bitreader hotfix

Commit:
methods changed to use blockynumbersaver instead of internal number writing system

Commit:
removed reporter features of blockydecompression

Commit:
fixed floatsimilardecompression

now the first batch of numbers gets decompression correctly, also
removed weird unique_ptr usage in blockydecompression

Commit:
fixed segmentation fault when accessing fill_buffer meta part multiple times in a row

Commit:
reworked fill_buffer decompress part

instead of filling an intermediary vector<char> with reconstructed BlockyNumbers, that are later pushed into the buffer (which was thoroughly stupid), BlockyNumbers are taken directly from the number saver and put into the buffer

Commit:
fixed bitreader EOF bug

also le funny printf after return statement xd

Commit:
fill_buffer now fills buffer with decomp. numbers

after having the numbersaver filled with actual reconstructed numbers
(which spawned an entirely new issue on its own), the numbers are now
emptied into the buffer correctly

Commit:
minor improvements

added a little more inline documentation for the fill_buffer method and some more return code shenanigans, also temporary segfault fix in blockydecompression #noshame

satorialist commented:
so recently, blocky decompression got quite a lot closer to completion. 

so much in fact, I dare say my solution is done and working. the code was only tested against alpha.water, so obviously that statement won't hold water. further testing definitely required.

there's two big issues (see issue #22 and #23) with blocky that are holding back the completion of blocky decompression, none of which are directly related to decompression

Commit:
finishing up before merging with SimuPoLAS/master

Commit:
merged SimuPoLAS/master

Commit:
fixed #23

Commit:
changed offset calculation for Section.START

knightmansmountain commented on src/blockydecompression.cpp:
```diff
@@ -6,156 +6,84 @@
+BlockyDecompression::BlockyDecompression(LZMAFILE* data, BlockyNumberSaver& saver, BitReader& reader, size_t size = 1)
+    : reader(reader), saver(saver), size(size)
 {
+    // TODO: get this to work
```
Works, can be deleted

UgiNyaa commented on src/bitreader.cpp:
```diff
@@ -16,16 +18,23 @@ BitReader::BitReader(FILE* file)
-        throw 0;
+    //auto c = getc(file);
+    char c;
+    lzmaread(&c, sizeof(char), 1, file);
```
You could check against the return type, which should represent the read count. If it does not equal 1, then an error probably occured or EOF happened. Thats at least the behaviour of the standard function fread: 
http://www.cplusplus.com/reference/cstdio/fread/

UgiNyaa commented on src/blockydecompression.cpp:
```diff
@@ -6,156 +6,84 @@
+    size_t value_count = 0;
+    // TODO: lmao this fix just stops the program on segaulting with alpha.water
+    // definitely temporary, remove ASAP
+    if (metadata.ValueCount > 2400) {
```
Isn't this already fixed? We did remove the error of having a faulty value count.

UgiNyaa commented on src/parsing2/decompressionparser.cpp:
```diff
@@ -0,0 +1,142 @@
+        // if the current_numbersave has no more numbers to be written to the buffer
+        if (!current_numbersaver.has_numbers()) {
+            // get new data
+            // TODO: put in a lot of work to get this to function correctly
```
This doesn't look very finished, can you evaluate?

UgiNyaa reviewed (commented):
It would also be cool if you finish the decompression of vectors and tensors. You can use the "U" file for testing. With that you are also able to test the size and RAM usage. However, you can also use "U.small" if RAM usage does not yet concern you.

knightmansmountain commented on src/io/blockystreambuffer.cpp:
```diff
@@ -82,16 +91,25 @@ BlockyStreamBuffer* BlockyStreamBuffer::close()
-    setg(buffer, buffer, buffer + num);
+    int processed = num;
+
+    // TODO: figure out a way to make the decompression parser say how much was
```
delete

knightmansmountain commented on src/methods/numbersnoexp/numbersnoexpdecompression.cpp:
```diff
@@ -23,7 +24,7 @@ int NumbersNoExpDecompression::read
 		BlockyNumber value = BlockyNumber(number, 0);
 		// AAAAAAAAA
```
Delete

knightmansmountain commented on src/parsing/hooker.cpp:
```diff
@@ -164,9 +165,16 @@ void Hooker::end_file()
+		off = section->End;
 		lzmawrite(&value, sizeof(int64_t), 1, meta);
 		lzmawrite(&section->Size, sizeof(uint8_t), 1, meta);
+		// TODO: REMOVE AFTER BUG IS FIXED
```
REMOVE

satorialist replied:
CHILL

knightmansmountain reviewed (approved):
Good job noby

Commit:
implemented suggested changes from #18
